### DISCUSSION TITLE
Delete orphan with attribute_mapped_collection causing "not in session"

### DISCUSSION THREAD

**Discussion by huntfx** [sqlalchemy/sqlalchemy#5996] [Usage Questions]
**Describe the bug**
I'm not really sure how to explain this well, so I'll try provide all the details of my particular use case.

I have the models `Info`, `Data` and `Link`, with `Link` acting as an association between the two. I'm attempting to get `attribute_mapped_collection` working properly with an `association_proxy`, so that I can end up using `info[key] = value` as valid syntax.

If I create 1 `Info` record and 2 `Data` records, then attempt to create 2 separate `Link` records with those, SQLalchemy will complain one of them is not in the session and will only commit one of the records.

I know it's a rare use case, but the error seems like a bug as in there's no obvious reason why it's "not in session".

**Expected behavior**
Both `Link` records should be inserted, or a more useful error message is shown why it's not happening.

These are the things I've found which will stop the issue happening
- Commit the session before adding the `Link` records
- Remove `delete-cascade` from `Info._link` (not ideal)
- Remove `attribute_mapped_collection` from `Info._link` (not ideal)

**To Reproduce**
```py
from sqlalchemy import Column, Integer, String, ForeignKey, create_engine
from sqlalchemy.orm import relationship, sessionmaker
from sqlalchemy.orm.collections import attribute_mapped_collection
from sqlalchemy.ext.associationproxy import association_proxy
from sqlalchemy.ext.declarative import declarative_base


Engine = create_engine('sqlite:///:memory:', echo=True)
Base = declarative_base(bind=Engine)
Session = sessionmaker(bind=Base.metadata.bind)


class Info(Base):
    __tablename__ = 'info'
    id = Column(Integer, primary_key=True)

    _link = relationship('Link', back_populates='info', collection_class=attribute_mapped_collection('key'), cascade='all, delete-orphan')
    data = association_proxy('_link', 'data', creator=lambda k, v: Link(data=Data(key=k, value=v)))


class Data(Base):
    __tablename__ = 'data'

    id = Column(Integer, primary_key=True)
    key = Column(String(32, collation='NOCASE'), nullable=False)
    value = Column(String(1024, collation='NOCASE'), nullable=False)

    _link = relationship('Link', back_populates='data', cascade='all, delete-orphan')
    info = association_proxy('_link', 'info')


class Link(Base):
    __tablename__ = 'link'
    info_id = Column(Integer, ForeignKey('info.id'), primary_key=True)
    data_id = Column(Integer, ForeignKey('data.id'), primary_key=True)

    data = relationship(Data, uselist=False, foreign_keys=data_id, back_populates='_link', innerjoin=True)
    info = relationship(Info, uselist=False, foreign_keys=info_id, back_populates='_link', innerjoin=True)
    key = association_proxy('data', 'key')


Base.metadata.create_all()


session = Session()
info = Info()
session.add(info)
data1 = Data(key='key', value='value')
session.add(data1)
data2 = Data(key='key2', value='value2')
session.add(data2)

link1 = Link(info=info, data=data1)
session.add(link1)
link2 = Link(info=info, data=data2)
session.add(link2)
session.commit()
```

**Error**
```
\site-packages\sqlalchemy\orm\unitofwork.py:284: SAWarning: Object of type <Link> not in session, add operation along 'Data._link' will not proceed
  % (orm_util.state_class_str(state), operation, prop)
2021-03-01 15:08:23,638 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2021-03-01 15:08:23,640 INFO sqlalchemy.engine.base.Engine INSERT INTO info DEFAULT VALUES
2021-03-01 15:08:23,641 INFO sqlalchemy.engine.base.Engine ()
2021-03-01 15:08:23,642 INFO sqlalchemy.engine.base.Engine INSERT INTO data ("key", value) VALUES (?, ?)
2021-03-01 15:08:23,642 INFO sqlalchemy.engine.base.Engine ('key', 'value')
2021-03-01 15:08:23,642 INFO sqlalchemy.engine.base.Engine INSERT INTO data ("key", value) VALUES (?, ?)
2021-03-01 15:08:23,643 INFO sqlalchemy.engine.base.Engine ('key2', 'value2')
2021-03-01 15:08:23,644 INFO sqlalchemy.engine.base.Engine INSERT INTO link (info_id, data_id) VALUES (?, ?)
2021-03-01 15:08:23,644 INFO sqlalchemy.engine.base.Engine (1, 2)
2021-03-01 15:08:23,645 INFO sqlalchemy.engine.base.Engine COMMIT
```

```py
>>> session.query(Link).all())  # Should be 2
1
```

**Versions.**
 - OS: Windows 10
 - Python: 3.7.9
 - SQLAlchemy: 1.3.13
 - Database: sqlite

**Comment by zzzeek** [comment]
Here's what I can see:

1. you are add()ing every object to the Session explcitly, so they should all be in there.
2. your output clearly states that one of those Link objects is *not* in the session:

> SAWarning: Object of type <Link> not in session, add operation along 'Data._link' will not proceed
>   % (orm_util.state_class_str(state), operation, prop)

3. the only way this happens is through those "delete-orphan" cascades.   They will kick an object out of the Session when they see it is being removed from their collection.
4. you have Link as the subject of two separate "delete-orphan" cascades, which is a problem.   What if Link is removed from its Info but not from its Data?  does it get deleted?   
5. Depending on which behavior you'd want, you'd set the [legacy_orphan](https://docs.sqlalchemy.org/en/13/orm/mapping_api.html?highlight=legacy_is_orphan#sqlalchemy.orm.mapper.params.legacy_is_orphan) flag for the "old" or "new" behavior.     but either way, it is not really a good idea to have two separate kinds of parent both able to kick your object out of the Session and I would recommend improving on that ambiguity.
6. well how did a Link get removed from a collection?  we're using a dictionary, ah.  so if the keys aren't correct, that is, in the very likely case that the key is None when the dictionary gets populated, the first Link will be replaced by the second Link and that's a delete-orphan
7. Why would the key be None?  because the "key" of Link is dependent on the "Data" object that *may or may not* be associated with it, and you are initializing as Link(info=info, data=data1).   I haven't tried it but I would hypothesize that Link is getting associated with "info" before "data", the key is None, goes in the dictionary with that key, on the second run the old one gets replaced.

Try it like this:

    link1 = Link(data=data1)
    assert link1.key is not None
    info.data[link1.key] = link1
